routing/turns_tts_text.cpp

```diff
@@ -62,20 +122,139 @@ std::string GetTtsText::GetTurnNotification(Notification const & notification) c
   if (notification.m_useThenInsteadOfDistance && notification.m_turnDir == CarDirection::None)
     return {};
 
-  std::string const dirStr = GetTextById(GetDirectionTextId(notification));
   if (dirStr.empty())
     return {};
 
-  std::string distStr;
+  std::string thenStr;
   if (notification.m_useThenInsteadOfDistance)
   {
-    distStr = GetTextById("then");
-    if (notification.m_distanceUnits > 0)
-      distStr = distStr + " " + GetTextById(GetDistanceTextId(notification));
+    // add "then" and space only if needed, for appropriate languages
+    thenStr = GetTextByIdTrimmed("then");
+    if (localeKey != "ja")
+      thenStr.push_back(' ');
+  }
+
+  std::string distStr;
+  if (notification.m_distanceUnits > 0)
+    distStr = GetTextByIdTrimmed(GetDistanceTextId(notification));
+
+  // Get a string like 245; CA 123; Highway 99; San Francisco
+  // In the future we could use the full RoadNameInfo struct to do some nice formatting.
+  std::string streetOut;
+  RouteSegment::RoadNameInfo nsi = notification.m_nextStreetInfo;  // extract non-const
+  FormatFullRoadName(nsi, streetOut);
+
+  if (!streetOut.empty())
+  {
+    // We're going to pronounce the street name.
+
+    // Replace any full-stop characters (in between sub-instructions) to make TTS flow better.
+    // Full stops are: . (Period) or 。 (East Asian) or । (Hindi)
+    RemoveLastDot(distStr);
+
+    // If the turn direction with the key +_street exists for this locale, and isn't "NULL",
+    // use it (like make_a_right_turn_street)
+    std::string dirStreetStr = GetTextByIdTrimmed(dirKey + "_street");
+    if (!dirStreetStr.empty() && dirStreetStr != "NULL")
+      dirStr = std::move(dirStreetStr);
+
+    // Normally use "onto" for "turn right onto Main St" unless it's "NULL"
+    std::string ontoStr = GetTextByIdTrimmed("onto");
+    if (ontoStr == "NULL")
+      ontoStr = "";
+
+    // If the nextStreetInfo has an exit number, we'll announce it
+    if (!notification.m_nextStreetInfo.m_junction_ref.empty())
+    {
+      // Try to get a specific "take exit #" phrase and its associated "onto" phrase (if any)
+      std::string dirExitStr = GetTextByIdTrimmed("take_exit_number");
+      if (!dirExitStr.empty())
+      {
+        dirStr = std::move(dirExitStr);
+        ontoStr.clear();  // take_exit_number overwrites "onto"
+      }
+    }
+
+    // Same as above but for dirStr instead of distStr
+    RemoveLastDot(dirStr);
+
+    std::string distDirOntoStreetStr = GetTextByIdTrimmed("dist_direction_onto_street");
+    // only load populate _street_verb if _street exists and isn't "NULL"
+    // may also need to handle a lack of a $5 position in the formatter string
+    std::string dirVerb = "";
```

Comment:
Strings are empty by default.
```suggestion
    std::string dirVerb;
```